### PR TITLE
Revision 2 of the ksql storage impl

### PR DESCRIPTION
**Pull request by EricWittmann**
Hey @Apicurio/developers - this is a PR of a second revision/evolution of the kafka+sql hybrid model we've been working on.  I think it captures most of what we documented in our design doc.  There are still some missing pieces, but I think it's ready for review so we can identify and itemize the remaining tasks.

### CODE REVIEW

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java [hunk 1]
```diff
@@ -253,125 +256,139 @@ private void startConsumerThread(final KafkaConsumer<String, Str.StorageValue> c
...
+            throw new ArtifactAlreadyExistsException(artifactId);
+        }
+        
+        String contentHash = ensureContent(content, artifactId, artifactType);
+        String createdBy = null; // TODO populate when authentication is available
+        Date createdOn = new Date();
+        
+        return submitter
+                .submitArtifact(artifactId, ActionType.Create, artifactType, contentHash, createdBy, createdOn, metaData)
+                .thenCompose(reqId -> (CompletionStage<ArtifactMetaDataDto>) coordinator.waitForResponse(reqId));
```
Comment (famartinrh) [inline storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java:314]: here we would catch if there is any error, and we would use log compaction to remove the specific kafka message for this action (using it's uuid)

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlSink.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    private Object processArtifactVersion(ArtifactVersionKey key, ArtifactVersionValue value) {
+        switch (value.getAction()) {
+            case Update:
+                sqlStore.updateArtifactVersionMetaDataAndState(key.getArtifactId(), key.getVersion(), value.getMetaData(), value.getState());
+                return null;
+            case Delete:
+                sqlStore.deleteArtifactVersion(key.getArtifactId(), key.getVersion());
+                return null;
+            case Clear:
+                sqlStore.deleteArtifactVersionMetaData(key.getArtifactId(), key.getVersion());
```
Comment (famartinrh) [inline storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlSink.java:196]: I was not aware we had overlapping concepts , having `delete` and `clear` is not perfect as the meaning is similar but I guess it's ok if we have no other option

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlValueSerializer.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     * @see org.apache.kafka.common.serialization.Serializer#serialize(java.lang.String, java.lang.Object)
+     */
+    @Override
+    public byte[] serialize(String topic, MessageValue messageValue) {
+        if (messageValue.getType() == MessageType.Content) {
+            return this.serializeContent(topic, (ContentValue) messageValue);
+        }
+        try (UnsynchronizedByteArrayOutputStream out = new UnsynchronizedByteArrayOutputStream()) {
+            out.write(ByteBuffer.allocate(1).put((byte) messageValue.getType().ordinal()).array());
+            mapper.writeValue(out, messageValue);
```
Comment (famartinrh) [inline storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlValueSerializer.java:54]: this is not protobuf nor avro :) but it's ok , we could even leave it like this but the message keys... that is another story

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlKeySerializer.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * @see org.apache.kafka.common.serialization.Serializer#serialize(java.lang.String, java.lang.Object)
+     */
+    @Override
+    public byte[] serialize(String topic, MessageKey messageKey) {
+        try {
+            UnsynchronizedByteArrayOutputStream out = new UnsynchronizedByteArrayOutputStream();
+            out.write(ByteBuffer.allocate(1).put((byte) messageKey.getType().ordinal()).array());
+            mapper.writeValue(out, messageKey);
```
Comment (famartinrh) [inline storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlKeySerializer.java:50]: using json for the message key does not look like a good idea to me

**Review by famartinrh — COMMENTED** [review]
I leaved some comments but overall looks really good, good job :)

File: app/src/main/java/io/apicurio/registry/storage/RegistryStorageException.java [hunk 5]
```diff
@@ -29,6 +29,7 @@
      */
     public RegistryStorageException(Throwable cause) {
         super(cause);
+        cause.printStackTrace();
```
Comment (carlesarnal) [inline app/src/main/java/io/apicurio/registry/storage/RegistryStorageException.java:32]: Why is this needed?

File: app/src/main/java/io/apicurio/registry/storage/RegistryStorageException.java [hunk 6]
```diff
@@ -38,6 +39,7 @@ public RegistryStorageException(Throwable cause) {
      */
     public RegistryStorageException(String reason, Throwable cause) {
         super(reason, cause);
+        cause.printStackTrace();
```
Comment (carlesarnal) [inline app/src/main/java/io/apicurio/registry/storage/RegistryStorageException.java:42]: Same as above

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/keys/MessageType.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package io.apicurio.registry.storage.impl.ksql.keys;
+
+/**
+ * @author eric.wittmann@gmail.com
+ */
+public enum MessageType {
+
+    GlobalRule, Content, Artifact, ArtifactRule, ArtifactVersion
```
[the remark this answers is not in the document]
Comment (carlesarnal) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/keys/MessageType.java:24]: +1

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlSink.java (hunk 2 above)
Comment (carlesarnal) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlSink.java:196]: What about joining both operations into a single one? Maybe I am missing something, but if the version is deleted, then the metadata can be safely deleted as well.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        return withHandle( handle -> {
+            String sql = sqlStatements().selectContentById();
+            return handle.createQuery(sql)
+                    .bind(0, contentId)
+                    .map(ContentMapper.instance)
+                    .one();
+        });
+    }
+
+    public CompletionStage<ArtifactMetaDataDto> createArtifactWithMetadata(String artifactId,
```
[the remark this answers is not in the document]
Comment (carlesarnal) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java:113]: The parent method that this method is calling is annotated. I don't think this is needed.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        if (metaData == null) {
+            ContentHandle content = this.getContent(contentId);
+            metaData = this.extractMetaData(artifactType, content);
+        }
+        
+        return super.createArtifactWithMetadata(artifactId, artifactType, contentId, createdBy, createdOn,
+                metaData, globalIdGenerator);
+    }
+    
+    public CompletionStage<ArtifactMetaDataDto> updateArtifactWithMetadata(String artifactId,
```
[the remark this answers is not in the document]
Comment (carlesarnal) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java:128]: This may be true here since here we're calling the protected method directly and that method doesn't have the annotation.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (carlesarnal) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java:113]: Sorry, you were right, here we're also calling the protected method without annotation.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/values/ActionType.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package io.apicurio.registry.storage.impl.ksql.values;
+
+/**
+ * @author eric.wittmann@gmail.com
+ */
+public enum ActionType {
+
+    Create, Update, Delete, Clear
```
[the remark this answers is not in the document]
Comment (carlesarnal) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/values/ActionType.java:24]: +1

**Review by carlesarnal — APPROVED** [review]
Besides the concern with the transactional methods, LGTM. Nice job!

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlStore.java:113]: *technically* this does not need to be transactional right now, because writes are single-threaded.  But still, it is more logically correct and consistent to make it transactional, so I will do that. :)

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlSink.java (hunk 2 above)
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/sql/KafkaSqlSink.java:196]: Yeah I didn't love this solution, but I didn't have a better one.  I'm open to other options.  The semantics here are actually:

1) Delete == delete the **artifact version**
2) Clear == delete the artifact version **metadata**

They are different actions associated with the same message key, which is important IMO for log compaction.  If you can think of either a better action name, or another approach, I'm happy to hear it.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlValueSerializer.java (hunk 3 above)
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlValueSerializer.java:54]: Yeah, my intention was to solicit a volunteer to re-implement the Serdes (both key and value) to use something like protobuf.  I felt like that could easily be done in a followup PR.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlKeySerializer.java (hunk 4 above)
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlKeySerializer.java:50]: Agree.  See previous comment.  :)  I didn't want to boil the ocean with this PR.  I have a list of improvements that need to be made.  I will create GitHub issues for them soon.

File: app/src/main/java/io/apicurio/registry/storage/RegistryStorageException.java (hunk 5 above)
Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/storage/RegistryStorageException.java:32]: Debugging leftover.  I will remove.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java [hunk 11]
```diff
@@ -253,125 +256,139 @@ private void startConsumerThread(final KafkaConsumer<String, Str.StorageValue> c
...
+        if (!sqlStore.isArtifactExists(artifactId)) {
+            throw new ArtifactNotFoundException(artifactId);
         }
 
+        String contentHash = ensureContent(content, artifactId, artifactType);
+        String createdBy = null; // TODO populate when authentication is available
+        Date createdOn = new Date();
+        
+        if (metaData == null) {
+            metaData = extractMetaData(artifactType, content);
```
[the remark this answers is not in the document]
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java:362]: I attacked Create and Update at different times and hilariously came up with different solutions.  In the Create case I'm re-loading the content from the DB in order to extract the metadata when processing the Kafka message.  In the Update case I'm extracting the metadata first and then sending it along as part of the Kafka message.  I think the latter is better so I will change the Create case to follow that pattern.

[the remark this answers is not in the document]
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java:362]: The net result of that change is we'll have more information in the Kafka message and thus less processing to do when consuming it.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java [hunk 12]
```diff
@@ -380,209 +397,324 @@ public ArtifactMetaDataDto getArtifactMetaData(String artifactId) throws Artifac
...
 
+    /**
+     * @see io.apicurio.registry.storage.RegistryStorage#deleteGlobalRules()
+     */
     @Override
     public void deleteGlobalRules() throws RegistryStorageException {
-        UUID reqId = ConcurrentUtil.get(submitter.submitRule(Str.ActionType.DELETE, GLOBAL_RULES_ID, null, null));
-        coordinator.waitForResponse(reqId);
-
+        // TODO find a better way to implement deleting all global rules?  This isn't very scalable if we add more rules.
```
[the remark this answers is not in the document]
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java:621]: This will be a followup GitHub issue.

[the remark this answers is not in the document]
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java:621]: How would we use log compaction for this?  We need multiple messages - one for each rule.  No?

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/keys/MessageType.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/keys/MessageType.java:24]: Agreed - I'll change this.  Also will change `ActionType` and do something for `ArtifactType` as well.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java (hunk 1 above)
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java:314]: Yes I will create a GH issue for this.

File: storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlValueSerializer.java (hunk 3 above)
Comment (EricWittmann) [reply storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/serde/KafkaSqlValueSerializer.java:54]: I will create a GH issue for this: https://github.com/Apicurio/apicurio-registry/issues/1059
